corehq/apps/es/forms.py

```diff
@@ -57,9 +69,100 @@ class ElasticForm(ElasticDocumentAdapter):
         return get_adapter_mapping(self)
 
-    @classmethod
-    def from_python(cls, doc):
-        return from_dict_with_possible_id(doc)
+    def from_python(self, xform):
+        from corehq.form_processor.models.forms import XFormInstance
+        if isinstance(xform, dict):
+            xform_dict = copy(xform)
+        elif isinstance(xform, XFormInstance):
+            xform_dict = xform.to_json()
+        else:
+            raise TypeError(f"Unknown type {type(xform)}")
+        return self._from_dict(xform_dict)
+
+    def _from_dict(cls, xform_dict):
+        """
+        Takes in a xform dict and applies required transformation to make it suitable for ES.
+        The function is replica of ``transform_form_for_elasticsearch``
+        and will be replaced by `from_python` in future
+
+        :param xform: an instance of ``dict`` which is ``XFormInstance.to_json()``
+        """
+        from casexml.apps.case.xform import extract_case_blocks
+
+        from corehq.apps.receiverwrapper.util import get_app_version_info
+        from corehq.pillows.utils import format_form_meta_for_es, get_user_type
+        from corehq.pillows.xform import is_valid_date
+
+        # create shallow copy of form object and case objects
+        # that will be modified in tranformation
+        form = xform_dict['form'] = copy(xform_dict['form'])
+
+        if 'case' in form:
+            if isinstance(form['case'], dict):
+                form['case'] = copy(form['case'])
+            elif isinstance(form['case'], list):
+                form['case'] = [copy(case) for case in form['case']]
+
+        user_id = None
+
+        if 'meta' in form:
+            form_meta = form['meta'] = copy(form['meta'])
+            if not is_valid_date(form_meta.get('timeEnd', None)):
+                form_meta['timeEnd'] = None
+            if not is_valid_date(form_meta.get('timeStart', None)):
+                form_meta['timeStart'] = None
+
+            # Some docs have their @xmlns and #text here
+            if isinstance(form_meta.get('appVersion'), dict):
+                form_meta = format_form_meta_for_es(form_meta)
+
+            app_version_info = get_app_version_info(
+                xform_dict['domain'],
+                xform_dict.get('build_id'),
+                xform_dict.get('version'),
+                form_meta,
+            )
+            form_meta['commcare_version'] = app_version_info.commcare_version
+            form_meta['app_build_version'] = app_version_info.build_version
+            user_id = form_meta.get('userID', None)
+
+            try:
+                geo_point = GeoPoint.from_string(xform_dict['form']['meta']['location'])
+                form_meta['geo_point'] = geo_point.lat_lon
+            except (KeyError, BadValueError):
+                form_meta['geo_point'] = None
+                pass
+
+        xform_dict['user_type'] = get_user_type(user_id)
+        xform_dict['inserted_at'] = datetime.utcnow().isoformat()
```

Comment:
I know you just copied this but it should probably use `json_format_datetime`

Reply:
https://github.com/dimagi/commcare-hq/pull/32660/commits/c660f9b6bd7f7f54eb22e8ad1e68bdec219ccaec